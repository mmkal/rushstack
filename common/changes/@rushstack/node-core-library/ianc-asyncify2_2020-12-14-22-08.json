{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "",
      "type": "none"
    }
  ],
  "packageName": "@rushstack/node-core-library",
  "email": "iclanton@users.noreply.github.com"
}