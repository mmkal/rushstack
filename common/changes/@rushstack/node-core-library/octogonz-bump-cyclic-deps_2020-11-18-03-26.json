{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "",
      "type": "none"
    }
  ],
  "packageName": "@rushstack/node-core-library",
  "email": "4673363+octogonz@users.noreply.github.com"
}