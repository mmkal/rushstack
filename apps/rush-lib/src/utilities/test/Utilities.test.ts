// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { StringBufferTerminalProvider, Terminal } from '@rushstack/node-core-library';
import { IDisposable } from 'rx';

import { Utilities } from '../Utilities';

describe('Utilities', () => {
  describe('printMessageInBox', () => {
    const MESSAGE: string =
      'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Maecenas porttitor congue massa. Fusce posuere, magna sed pulvinar ultricies, purus lectus malesuada libero, sit amet commodo magna eros quis urna.';

    let terminalProvider: StringBufferTerminalProvider;
    let terminal: Terminal;

    beforeEach(() => {
      terminalProvider = new StringBufferTerminalProvider(false);
      terminal = new Terminal(terminalProvider);
    });

    afterEach(() => {
      jest.resetAllMocks();
    });

    function validateOutput(expectedWidth: number): void {
      const outputLines: string[] = terminalProvider
        .getOutput({ normalizeSpecialCharacters: true })
        .split('[n]');
      expect(outputLines).toMatchSnapshot();

      expect(outputLines[0].trim().length).toEqual(expectedWidth);
    }

    it('Correctly prints a narrow box', () => {
      Utilities.printMessageInBox(MESSAGE, terminal, 20);
      validateOutput(20);
    });

    it('Correctly prints a wide box', () => {
      Utilities.printMessageInBox(MESSAGE, terminal, 300);
      validateOutput(300);
    });

    it('Correctly gets the console width', () => {
      jest.spyOn(Utilities, 'getConsoleWidth').mockReturnValue(65);

      Utilities.printMessageInBox(MESSAGE, terminal);
      validateOutput(32);
    });
  });

  describe('usingAsync', () => {
    let disposed: boolean;

    beforeEach(() => {
      disposed = false;
    });

    class Disposable implements IDisposable {
      public dispose(): void {
        disposed = true;
      }
    }

    it('Disposes correctly in a simple case', async () => {
      await Utilities.usingAsync(
        () => new Disposable(),
        () => {
          /* no-op */
        }
      );

      expect(disposed).toEqual(true);
    });

    it('Disposes correctly after the operation throws an exception', async () => {
      await expect(
        async () =>
          await Utilities.usingAsync(
            () => new Disposable(),
            () => {
              throw new Error('operation threw');
            }
          )
      ).rejects.toMatchSnapshot();

      expect(disposed).toEqual(true);
    });

    it('Does not dispose if the construction throws an exception', async () => {
      await expect(
        async () =>
          await Utilities.usingAsync(
            async () => {
              throw new Error('constructor threw');
            },
            () => {
              /* no-op */
            }
          )
      ).rejects.toMatchSnapshot();

      expect(disposed).toEqual(false);
    });
  });
});
