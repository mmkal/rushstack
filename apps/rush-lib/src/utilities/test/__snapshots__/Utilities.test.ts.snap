// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utilities printMessageInBox Correctly gets the console width 1`] = `
Array [
  " ╔══════════════════════════════╗ ",
  " ║    Lorem ipsum dolor sit     ║ ",
  " ║      amet, consectetuer      ║ ",
  " ║       adipiscing elit.       ║ ",
  " ║      Maecenas porttitor      ║ ",
  " ║     congue massa. Fusce      ║ ",
  " ║      posuere, magna sed      ║ ",
  " ║     pulvinar ultricies,      ║ ",
  " ║         purus lectus         ║ ",
  " ║    malesuada libero, sit     ║ ",
  " ║      amet commodo magna      ║ ",
  " ║       eros quis urna.        ║ ",
  " ╚══════════════════════════════╝ ",
  "",
]
`;

exports[`Utilities printMessageInBox Correctly prints a narrow box 1`] = `
Array [
  " ╔══════════════════╗ ",
  " ║      Lorem       ║ ",
  " ║      ipsum       ║ ",
  " ║    dolor sit     ║ ",
  " ║      amet,       ║ ",
  " ║   consectetuer   ║ ",
  " ║    adipiscing    ║ ",
  " ║      elit.       ║ ",
  " ║     Maecenas     ║ ",
  " ║    porttitor     ║ ",
  " ║      congue      ║ ",
  " ║      massa.      ║ ",
  " ║      Fusce       ║ ",
  " ║     posuere,     ║ ",
  " ║    magna sed     ║ ",
  " ║     pulvinar     ║ ",
  " ║    ultricies,    ║ ",
  " ║      purus       ║ ",
  " ║      lectus      ║ ",
  " ║    malesuada     ║ ",
  " ║     libero,      ║ ",
  " ║     sit amet     ║ ",
  " ║     commodo      ║ ",
  " ║      magna       ║ ",
  " ║    eros quis     ║ ",
  " ║      urna.       ║ ",
  " ╚══════════════════╝ ",
  "",
]
`;

exports[`Utilities printMessageInBox Correctly prints a wide box 1`] = `
Array [
  " ╔══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╗ ",
  " ║                                              Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Maecenas porttitor congue massa. Fusce posuere, magna sed pulvinar ultricies, purus lectus malesuada libero, sit amet commodo magna eros quis urna.                                               ║ ",
  " ╚══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╝ ",
  "",
]
`;

exports[`Utilities usingAsync Disposes correctly after the operation throws an exception 1`] = `[Error: operation threw]`;

exports[`Utilities usingAsync Does not dispose if the construction throws an exception 1`] = `[Error: constructor threw]`;
